<div class="namebox">    
Double Click to edit and enter your

1.  Aparna Mohan
2.  862396080
3.  amoha121@ucr.edu
    
</div>

In [4]:
!cs203 job run ' getconf -a | grep CACHE'

srun -N1 -p cs203  getconf -a | grep CACHE
LEVEL1_ICACHE_SIZE                 32768
LEVEL1_ICACHE_ASSOC                8
LEVEL1_ICACHE_LINESIZE             64
LEVEL1_DCACHE_SIZE                 32768
LEVEL1_DCACHE_ASSOC                8
LEVEL1_DCACHE_LINESIZE             64
LEVEL2_CACHE_SIZE                  524288
LEVEL2_CACHE_ASSOC                 8
LEVEL2_CACHE_LINESIZE              64
LEVEL3_CACHE_SIZE                  16777216
LEVEL3_CACHE_ASSOC                 0
LEVEL3_CACHE_LINESIZE              64
LEVEL4_CACHE_SIZE                  0
LEVEL4_CACHE_ASSOC                 0
LEVEL4_CACHE_LINESIZE              0


<div class="alert alert-block alert-success"><div class="question correctness">

### Question 1 (Correctness)

<div class="question-text">

Reading the output from the above command, let's focus on "LEVEL1_D", namely, the L1 data cache configuration first. Remember the $C=ABS$ equation we taught in class, can you try to answer the following questions? 
</div>
    
<div class="answer">

**What's the value of C:** 
    
**What's the value of A:**

**What's the value of B:**

**What's the value of S:**

**How many bits are there in the offset?:**

**How many bits are there in the index?:**
    
</div>
</div>
</div>

<div class="alert alert-block alert-success">
<div class="answer">
You answers:
    
**What's the value of C:** C stands for Capacity, the capacity of L1 Cache is 32768
    
**What's the value of A:** A stands for Associative. From the above table L1 Cache has A=8

**What's the value of B:** B stands for Blocksize (Cacheline). Here B = 64

**What's the value of S:** For L1 Cache, S = C/A*B, S = 64

**How many bits are there in the offset?:** Offset = lg(B) = lg(64) = 6 bits

**How many bits are there in the index?:** Index = lg(S) = 6 bits
    
</div>
</div>

<div class="alert alert-block alert-success"><div class="question correctness">

### Question 2 (Correctness)

<div class="question-text">
Given the cache configuration we've learned in Question 1. If `size == 8192`, how many total L1 data cache misses do think will occur during the execution of `array()`? Assume the cache is empty to begin and you may assume the `data` address starts with something like 0x40000 as we did in the lecture. 
</div>
    
<div class="answer">

**Start with filling the table with the first 16 memory accesses**
| data[?] | address | tag | index | offset| hit or miss|
|---------|---------|-----|-------|-------|------------|
| data[0] | 0x40000 | 0x40|    0x0|    0x0| miss       |
|         |         |     |       |       |            |
|         |         |     |       |       |            |
|         |         |     |       |       |            |
|         |         |     |       |       |            |
|         |         |     |       |       |            |
|         |         |     |       |       |            |
|         |         |     |       |       |            |
|         |         |     |       |       |            |
|         |         |     |       |       |            |
|         |         |     |       |       |            |
|         |         |     |       |       |            |
|         |         |     |       |       |            |
|         |         |     |       |       |            |
|         |         |     |       |       |            |
|         |         |     |       |       |            |

**What's the estimate cache miss rate? Please explain. You may find LaTeX formula syntax useful.**
</div>
</div>
</div>

<div class="alert alert-block alert-success">
    
<div class="answer">

**Start with filling the table with the first 16 memory accesses**
    
| data[?] | address | tag | index | offset| hit or miss|
|---------|---------|-----|-------|-------|------------|
| data[0] | 0x40000 | 0x40| 0x0   | 0x0   | miss       |
| data[1] | 0x40008 | 0x40| 0x0   | 0x8   | hit        |
| data[2] | 0x40010 | 0x40| 0x0   | 0x10  | hit        |
| data[3] | 0x40018 | 0x40| 0x0   | 0x18  | hit        |
| data[4] | 0x40020 | 0x40| 0x0   | 0x20  | hit        |
| data[5] | 0x40028 | 0x40| 0x0   | 0x28  | hit        |
| data[6] | 0x40030 | 0x40| 0x0   | 0x30  | hit        |
| data[7] | 0x40038 | 0x40| 0x0   | 0x38  | hit        |
| data[8] | 0x40040 | 0x40| 0x1   | 0x0   | miss       |
| data[9] | 0x40048 | 0x40| 0x1   | 0x8   | hit        |
| data[10]| 0x40050 | 0x40| 0x1   | 0x10  | hit        |
| data[11]| 0x40058 | 0x40| 0x1   | 0x18  | hit        |
| data[12]| 0x40060 | 0x40| 0x1   | 0x20  | hit        |
| data[13]| 0x40068 | 0x40| 0x1   | 0x28  | hit        |
| data[14]| 0x40070 | 0x40| 0x1   | 0x30  | hit        |
| data[15]| 0x40078 | 0x40| 0x1   | 0x38  | hit        |

**What's the estimate cache miss rate? Please explain. You may find LaTeX formula syntax useful.**
</div>
</div>

<div class="alert alert-block alert-success"><div class="question completeness"> 

### Question 3 (Completeness)

<div class="question-text">

Predict how many bytes of the structs below will occupy.
    
</div>
   
<div class="answer">

| struct | sizeof() |
|--------|----------|
| `struct_1` |     | 
| `struct_2` |     | 
| `struct_3` |     | 
| `struct_4` |     | 
| `struct_5` |     | 
| `struct_6` |     | 
| `struct_7` |     | 
| `struct_8` |     | 
    
</div>
</div>
</div>

<div class="alert alert-block alert-success">    
<div class="answer">

    
| struct     | sizeof() |
|--------    |----------|
| `struct_1` |   4      | 
| `struct_2` |   8      | 
| `struct_3` |   8      | 
| `struct_4` |   12     | 
| `struct_5` |   4      | 
| `struct_6` |   12     | 
| `struct_7` |   16     | 
| `struct_8` |   16     | 
    
</div>
</div>

<div class="alert alert-block alert-success"><div class="question optional">

### Question 4 (Optional)
<div class="question-text">

How would you implement `lat_mem_rd`?
    
</div>
</div>
</div>

<div class="alert alert-block alert-success">
    
<div class="answer">
Answer:
    
How would you implement `lat_mem_rd`?
    
</div>
</div>

<div class="alert alert-block alert-success"><div class="question correctness points-4">

### Question 5 (Correctness)

<div class="question-text">

Based on this graph, give the latencies for each of the caches.
</div>
</div>
</div>

<div class="alert alert-block alert-success">
<div class="answer">

| Level of the memory hierarchy | latency (ns) |  latency (Cycles) |  
| ------------------------------|--------------|-------------------| 
| L1                            |0.87          |         4         |
| L2                            |2.62          |         10        |
| L3                            |3.029         |         12        |
| Main memory                   |5             |         19        |
    
</div>
    
</div>

<div class="alert alert-block alert-success">
<div class="question correctness points-8"> 

### Question 6 (Correctness)

<div class="question-text">

Given the following `tensor_t` declarations and accesses, compute the total size of the tensor in element and bytes, and the index (starting at 0) of the corresponding data elements in the `data` array.
    
</div>
   
<div class="answer">

1.  `tensor_t<uint32_t>(3,5,6,7)`
    1. Total elements:  
    2. Total bytes: 
    3. `get(1,0,0,0)`: 
    4. `get(0,1,0,0)`: 
    5. `get(0,0,1,0)`: 
    6. `get(0,0,0,1)`: 
2.  `tensor_t<double>(2,4,8,16)`
    1. Total elements:  
    2. Total bytes: 
    3. `get(1,3,2,4)`: 
    4. `get(2,2,1,7)`: 
            
</div>
</div>
</div>

<div class="alert alert-block alert-success">   
<div class="answer">

1.  `tensor_t<uint32_t>(3,5,6,7)`
    1. Total elements:  
    Given data : 4-dimensional tensor
    To find the total elements, I'm computing the product 3*5*6*7 = 630  
    2. Total bytes: 
    To find the total bytes I'm computing the product of total elements and uint32_t = 630 * 4 = 2520
    3. `get(1,0,0,0)`: 
    Considering the given value tensor_t<uint32_t>(3,5,6,7):
    The solution for get(1,0,0,0) is 1.
    4. `get(0,1,0,0)`: 1 * 3 = 3
    5. `get(0,0,1,0)`: 1 * 3 * 5 = 15
    6. `get(0,0,0,1)`: 1 * 3 * 5 * 6 = 90
2.  `tensor_t<double>(2,4,8,16)`
    1. Total elements:  
    Given data : 4-dimensional tensor
    To find the total elements, I'm computing the product 2*4*8*16 = 1024  
    2. Total bytes: 
    To find the total bytes I'm computing the product of total elements and size of double = 1024 * 8 = 8192
    3. `get(1,3,2,4)`: 4*(2*4*8) + 2*(2*4) + 3*2 + 1 = 256 + 16 + 6 + 1 = 279
    4. `get(2,2,1,7)`: 7*64 + 1*8 + 2*2 + 2 = 448 + 8 + 4 + 2 = 462
            
</div>
</div>

<div class="alert alert-block alert-success"><div class="question correctness">

### Question 7 (Correctness)

<div class="question-text">
How many L1 D-cache misses would you expect for `stride == 1`?  How about `stride == 8` and `stride == 16`?  Assume `size` is very large and that cache lines are 64 bytes. 
</div>
<div class="answer">

**`stride == 1` L1 D-Misses:** 

**`stride == 8` L1 D-Misses:**
    
**`stride == 16` L1 D-Misses:** 

</div>
</div>
</div>

<div class="alert alert-block alert-success">
<div class="answer">

**`stride == 1` L1 D-Misses:** 
    <br>Given data: size of the block is 64 bytes, size(uint32_t) is 4.
    <br>Cache miss = Size of the block/size(uint32_t) = 64/4 = 16.
    <br>Since it is given that size is very large temporal locality is disregarded.
    <br>**So, L1 D-Misses for stride==1 is size/16**

**`stride == 8` L1 D-Misses:**
    <br>Given data: size of the block is 64 bytes, stride*size(uint32_t) = 8*4 = 32.
    <br>Cache miss = Size of the block/32 = 64/32 = 2.
    <br>Since it is given that size is very large temporal locality is disregarded.
    <br>**So, L1 D-Misses for stride==8 is size/2**
    
    
**`stride == 16` L1 D-Misses:** 
    <br>Given data: size of the block is 64 bytes, stride*size(uint32_t) = 16*4 = 64.
    <br>Cache miss = Size of the block/64 = 64/64 = 1.
    <br>Since it is given that size is very large temporal locality is disregarded.
    <br>**So, L1 D-Misses for stride==8 is size**
    

</div>
</div>
</div>

<div class="alert alert-block alert-success">
<div class="question completeness">

### Question 8 (Completeness)

<div class="question-text">
Do the measurements match our predictions?  Does anything seem surprising about the results?
    </div>
</div>
</div>

<div class="alert alert-block alert-success">
<div class="answer">
Answer: From the graph above and on comparison with the generated csv, I can observe that L1D cache misses are higher than what I expected.
</div>
</div>

<div class="alert alert-block alert-success"><div class="question correctness points-2">

### Question 9 (Correctness)

<div class="question-text">
    
If a cache miss increases memory instruction latency by 200 cycles, how much should CPI increase between a stride of 1 and a stride of 64 (based on the data above)?
    
</div>
</div>
</div>

<div class="alert alert-block alert-success">

<div class="answer">

**Your estimation**

**Show your work**
    
<br> Average CPI = hitcycle + (cache missrate) ∗ (100% ∗ main memory cycle + percentage of memory instructions ∗ main memory cycle) 

<br> Average CPI = hitcycle + (cache missrate) ∗ (100% ∗ main memory cycle + (No of memory instruction/IC) ∗ main memory cycle) 
    
<br> Average CPI = hitcycle + (cache missrate) ∗ (100% ∗ main memory cycle + (L1 dcache accesses/IC) ∗ main memory cycle) 

<br>When CPI stride is 1 = 4 + 0.03 ∗ (200 + (3354943886/12815619926)∗ 200) = 11.57 

<br>When CPI stride is 16 = 4 + 0.42 ∗ (200 + (3359088761/12824202531) ∗ 200) = 109.84 

<br>From this, we can conclude that if CPI is more by 9.4x (109/84/11.57)when stride of 16 is considered over the stride of 1.
    
</div>
</div>
</div>

<div class="alert alert-block alert-success"><div class="question correctness">

### Question 10 (Correctness)
<div class="question-text">
    
Given the conditions described above, estimate the _number of cache misses_ that will occur for `size = 1024`, `size = 4096`, and `size = 16384`.  Assume we run `stride()` 10,000 times with the same values of `data` and `size`. The stride size is 16.

</div>
</div>
</div>

<div class="alert alert-block alert-success">    
<div class="answer">
    
**Cache misses for size = 1024:**
    <br> Given data: Stride 16, size is 1024
    <br> Size/stride = 1024/16 = 64
    <br> Conclusion : Compulsory miss occurs but after that all memory access will be a hit. **64 Cache misses can occur**. 
 
**Cache misses for size = 4096:**
    <br> Given data: Stride 16, size is 4096
    <br> Size/stride = 4096/16 = 256
    <br> Conclusion : Compulsory miss occurs but after that all memory access will be a hit. **256 Cache misses can occur**. 

**Cache misses for size = 16384:**
    <br> Given data: Stride 16, size is 16384
    <br> It can be observed that temporal and spatial locality can be disregarded as it is also mentioned that there is very little spatial locality. The inner for loop iterates 1024. 
    <br> Cache miss =  1024 * 10000(program is run so many times) * 16 (the outer for loop) = 163840000
**Show your work here**
</div>
</div>

<div class="alert alert-block alert-success"><div class="question correctness">

### Question 11 (Correctness)<div class="question-text">
    
What's the major types of misses in Question 10 when size == 1024 and size == 16384. 

</div>
</div>
</div>

<div class="alert alert-block alert-success">
<div class="answer">

**What's the major type of 'stride()' in Question 10 when `size == 1024`?**
    When size is 1024, compulsory misses occur.

**What's the major type of 'stride()' in Question 10 when `size == 16384`?**
    When size is 16384, capacity misses occur.

</div>
</div>

<div class="alert alert-block alert-success"><div class="question completeness">

### Question 12 (Completeness)<div class="question-text">
    
Assume our 32kB cache with 64-byte lines and 8-way associativity and 64-bit addresses.  Given an address $A$, how can we compute a new address, $B$, that will map to the same associative set but is not part of the same cache line as $A$?  Given an index, $i$, into an array, how can we compute the index of another element, $j$, that will conflict with the first?
    
    C = ABS
    Given data C = 32KB , A =8 ,  64 bit addresses , new address = ?
    Address = 52 (tag) + 6 (tag) + 6 (offset)
    On plugging in values, I get 12 bits that is 4096, which is the size to change tag.
    
    

</div>
</div>
</div>

<div class="alert alert-block alert-success">
<div class="answer">

**How do you compute `B`?**
To compute B, B = A + 0x1000 (12 bit)

**How do you compute `j`?**
To compute j, considering data type as uint32_t j = 0x1000/4 = 1024
</div>
</div>

<div class="alert alert-block alert-success"><div class="question completeness">
    
### Question 13 (Completeness)<div class="question-text">
    
Based on our analysis above, what do you think will happen with if the stride is one cache line longer (1040 bytes) or or one cache line shorter (1008 bytes)?  Why?
    
</div>
</div>
</div>

<div class="alert alert-block alert-success">    
<div class="answer">
    
**Stride 1008:**
   <br> For Stride 1008, I feel that the performance will be enhanced and will reach sets of multiple associativity.
**Stride 1040:**
    <br> For Stride 1040, I feel that the performance will be enhanced and will reach sets of multiple associativity.
</div> 
</div>

<div class="alert alert-block alert-success"><div class="question optional">

### Question 14 (Optional)<div class="question-text">

Consider the implementation of `tensor_t` described earlier in the assignment.  Accessing a tensor column-wise produces strided accesses which could lead to conflict misses if the dimensions of the tensor are "unlucky".  Why is this so?  What constitutes "unlucky" dimensions?  How could you modify `tensor_t` make it (mostly) immune to "unlucky" dimensions?
    
</div>
</div>
</div>
 

   
<div class="alert alert-block alert-success">
<div class="answer">
Answer: 
</div>
</div>

<div class="alert alert-block alert-success"><div class="question completeness">
    
### Question 15 (Completeness)<div class="question-text">
    
As `size` increases, the miss rate for the L1 and L2 will rise.  At value of `size` would you expect to see significant increases in L1 and L2 miss rates? Why?
</div>
</div>
</div>

<div class="alert alert-block alert-success">
<div class="answer">

**L1 critical `size`:**
    L1's critical size is computed as follows: Dcache size/ uint32_t size

**L2 critical `size`:**
    L2's critical size is computed as follows: Dcache size/ uint32_t size
    
</div>
</div>

<div class="alert alert-block alert-success"><div class="question completeness">
    
### Question 16 (Completeness)
    
<div class="question-text">
    
Using the code above, what values of `BYTES` and `size` should we run the miss machine with to measure the L1-TLB miss latency?  (The fact that there are two experiments listed is a hint that you'll need to run two different experiments.)
    
</div>
</div>
</div>    

<div class="alert alert-block alert-success">   
<div class="answer">

|| `BYTES` | `size` |
|--|------------|--|
|Experiment 1|4160 | 4160|
|Experiment 2|4160 | 2129920|
    
</div>
</div>

<div class="alert alert-block alert-success"><div class="question optional">
    
### Question 17 (Optional)

<div class="question-text">

The measurement above is for a miss to the L1 TLB.  Perform a different experiment to measure the L2 TLB miss latency.  This is harder than it appears at first.
    
</div>
</div>    
</div>

    
<div class="alert alert-block alert-success">    
<div class="answer">
Answer:
    
</div>    
</div>

<div class="alert alert-block alert-success"><div class="question optional">
    
### Question 18 (Optional)
    
<div class="question-text">

The measurements above are based on 4kB pages, but we can also use 2MB "huge pages".  Repeat the experiment above to determine whether 2MB TLB entries can also reside in the L2 TLB.      
</div>
<div class="answer">

A few notes:
    
1.  This one is a little involved.  You'll need to significantly tweak the experiments we did above.
2.  Whether 2MB TLB entries can be in the in the L2 TLB is not clearly specified in any documents I have found, so I don't know the answer.
3.  To get the system to use 2MB huge pages, remove the call to `madvise()` in `TLB.cpp` and ask `posix_memalign()` to give 2MB-aligned memory.
4.  Look in `TLB.cpp` for examples of how to change `BYTES`.  `TLB_2M()` is a good starting point.

</div>
</div>
</div>

<div class="alert alert-block alert-success">
<div class="answer">

A few notes:
    
1.  This one is a little involved.  You'll need to significantly tweak the experiments we did above.
2.  Whether 2MB TLB entries can be in the in the L2 TLB is not clearly specified in any documents I have found, so I don't know the answer.
3.  To get the system to use 2MB huge pages, remove the call to `madvise()` in `TLB.cpp` and ask `posix_memalign()` to give 2MB-aligned memory.
4.  Look in `TLB.cpp` for examples of how to change `BYTES`.  `TLB_2M()` is a good starting point.

</div>
</div>
    

In [ ]:
!git config --global user.email "you@example.com" !git config --global user.name\ 
"Your Name"